MRTK vNext - Open vr device initial implementation

SimonDarksideJ opened the pull request:
Overview
---
First pass for the OpenVR device implementation

Generic Headset and controllers, tested using Windows MR on a standalone platform using SteamVR

Changes
---
- [X] Added OpenVR Device manager
- [X] Added GenericOpenVRController
- [X] Controller Positional tracking (no grip available)
- [X] Controller Button actions (except Grip, as OpenVR reports it as a SingleAxis, not button)
- [X] Controller Axis input
- [X] Other controller types (Although, no mechanism to detect type yet. Will review in another PR)
- [ ] Test with more devices (on the delivery van)
- [X] walk on water
- [x] Curse Unity (because you cannot tell the difference between left hands, if you have more than one)
- [X] Added new utility to manage Unity Input Mapping management

Notes
---
Just in initial phase, I'm sharing as I have my controller visualizer showing cubes moving :D 
Controller fully tested and working using SteamVR with WMR.  Need more SDK testing to finish.
**Remaining outstanding work will be done in new Tasks.**

Base implementation ready for review.

Jarodshow commented on Assets/MixedRealityToolkit/_Core/Devices/OpenVR/GenericOpenVRController.cs:
```diff
@@ -12,5 +16,335 @@ public class GenericOpenVRController : BaseController
+
+        private Vector3 currentPointerPosition = Vector3.zero;
+        private Quaternion currentPointerRotation = Quaternion.identity;
+        private MixedRealityPose currentPointerData = new MixedRealityPose(Vector3.zero, Quaternion.identity);
```
`MixedRealityPose.ZeroIdentity`?

SimonDarksideJ replied:
Updated. Will need a separate PR to update the WMRcontrollers this was copied from

StephenHodgson replied:
I had thought I addressed this already in the WMR class

SimonDarksideJ replied:
Just checked and we had, I must have simply copied it before that change :D

StephenHodgson commented on Assets/MixedRealityToolkit/_Core/Devices/OpenVR/OpenVRDeviceManager.cs:
```diff
@@ -1,11 +1,158 @@
+            if (IsNodeTypeSupported(obj))
+            {
+                var controller = GetOrAddController(obj);
+                InputSystem?.RaiseSourceDetected(controller?.InputSource, controller);
```
Why would a controller be null here?

SimonDarksideJ replied:
Belt and braces. Copied from the WMR implementation.

StephenHodgson replied:
Let's raise an error or assert that it's not null.

There's not really any reason why it would be null here.

SimonDarksideJ replied:
Fair enough.   But will do a separate PR to address both at the same time so we don't get confused :D

StephenHodgson commented on Assets/MixedRealityToolkit/_Core/Managers/MixedRealityManager.cs:
```diff
@@ -166,9 +167,30 @@ private void Initialize()
+            {
+                case RuntimePlatform.WindowsPlayer:
+                case RuntimePlatform.WindowsEditor:
+                    AddManager(typeof(IMixedRealityDeviceManager), new OpenVRDeviceManager("OpenVR Device Manager", 10));
```
What if I wanted to use an open vr controller for wsa?

SimonDarksideJ replied:
Now that is a subject for later.  Not even going to ponder that until my test hardware arrives :D

StephenHodgson replied:
I'm also wondering if the WMR controllers show up in the openvr APIs

SimonDarksideJ replied:
That I can answer, no.  Or at least I don't think so.
The non WSA XR stuff only "lights up" if the VR platform is set to "OpenVR".
In UWP it's set to "Windows Mixed Reality".  But it doesn't hurt, to test, LATER!!

StephenHodgson replied:
So then in reality we shouldn't be checking the build platform, but which SDK is currently loaded in?

SimonDarksideJ replied:
Temporary measure while I add the OpenVR platform.  Will review SDK detection in a separate PR, as that is not the focus of this PR :D

keveleigh replied:
> I'm also wondering if the WMR controllers show up in the openvr APIs

They might show up if you're using the [WMR/Steam utility](https://store.steampowered.com/app/719950/Windows_Mixed_Reality_for_SteamVR/).

SimonDarksideJ replied:
That's how I'm testing this at the moment :D 
Using the SteamVR wrapper in Unity standalone with WMR.  I have other test kit "on the way" to test everything with the other SDK's

SimonDarksideJ commented:
Please ignore the Debug statements in the latest commit, they are for testing only.

And if anyone has found a way to get Input.Axis data WITHOUT the damn InputMapping, please comment!
And no, the ```Input.GetAxis("axis 1")``` code in there DOES NOT work :S #grumble

StephenHodgson commented on Assets/MixedRealityToolkit/_Core/Devices/OpenVR/GenericOpenVRController.cs:
```diff
@@ -138,211 +138,237 @@ private void UpdateControllerData(XRNodeState State, MixedRealityInteractionMapp
+                case DeviceInputType.Trigger:
+                    {
+                        //Get the current Trigger axis state - ** Does not WORK
+                        var triggerAxis = ControllerHandedness == Handedness.Left ? Input.GetAxis("axis 9") : Input.GetAxis("axis 10");
```
Not sure about having to do this by Axis. We can get buttons fine, but axes are still something users have to setup by hand, which can be annoying.

SimonDarksideJ replied:
This was a rage check-in as the above does not work.  As per the updated descripting, I'm really hating Unity input right now.

StephenHodgson commented:
I wanna dig into it and see if we can get the input directly or something along those lines.

davidkline-ms commented:
>wanna dig into it and see if we can get the input directly or something along those lines.
I don't wanna have to rely on Unity's old Input system, and the new one is still a ways off.

Does this imply platform specific input code for EVERY platform we support?

SimonDarksideJ commented:
> Does this imply platform specific input code for EVERY platform we support?

Not really @davidkline-ms , the default is Unity input system and where possible (like with WMR eventually) we'll use platform targeted code.  
It's just that Input.GetAxis FORCES you to use the input  mapping.  Everything else is possible with just code.

SimonDarksideJ commented:
Going back through the docs, I'll need to do another pass to add new events for things we didn't have in WMR, e.g. Trigger touch.

StephenHodgson commented on Assets/MixedRealityToolkit/_Core/Utilities/InputMappingAxisUtility.cs:
```diff
@@ -0,0 +1,356 @@
+        public const string CONTROLLER_LEFT_TRIGGER = "CONTROLLER_LEFT_TRIGGER";
+        public const string CONTROLLER_RIGHT_TRIGGER = "CONTROLLER_RIGHT_TRIGGER";
+        public const string XBOX_SHARED_TRIGGER = "XBOX_SHARED_TRIGGER";
+        public const string XBOX_A = "XBOX_A";
```
We don't need to map the buttons, only axes

StephenHodgson commented on Assets/MixedRealityToolkit/_Core/Utilities/InputMappingAxisUtility.cs:
```diff
@@ -0,0 +1,356 @@
+    /// Input from all types should be defined here for use 
+    /// though out the entire toolkit.
+    /// </summary>
+    public static class InputMappingAxisUtility
```
I'm not sure how I feel about using the old input system. I was really hoping to get the input directly from the SDK or driver.

SimonDarksideJ replied:
Removed the Xbox defaults, as for the InputManager. These work for now until we come up with a better solution later. Start from what works and then iterate.

StephenHodgson commented on Assets/MixedRealityToolkit/_Core/Utilities/InputMappingAxisUtility.cs:
```diff
@@ -0,0 +1,356 @@
+        public const string CONTROLLER_RIGHT_STICK_HORIZONTAL = "CONTROLLER_RIGHT_STICK_HORIZONTAL";
+        public const string CONTROLLER_RIGHT_STICK_VERTICAL = "CONTROLLER_RIGHT_STICK_VERTICAL";
+        public const string XBOX_DPAD_HORIZONTAL = "XBOX_DPAD_HORIZONTAL";
+        public const string XBOX_DPAD_VERTICAL = "XBOX_DPAD_VERTICAL";
```
Let's not add the Xbox controller Mappings, we won't need them where we're going.

Remember, we're getting the input from the windows namespace directly via `Windows.Gaming`

SimonDarksideJ replied:
Fair enough, I'll remove the Xbox controller "stuff"

StephenHodgson commented on Assets/MixedRealityToolkit/_Core/Utilities/InputMappingAxisUtility.cs:
```diff
@@ -0,0 +1,249 @@
+        /// </summary>
+        public class InputManagerAxis
+        {
+            public string Name = "";
```
`string.Empty` please

SimonDarksideJ replied:
Dang, that'll be from the old code.  Good catch

StephenHodgson commented on Assets/MixedRealityToolkit/_Core/Devices/OpenVR/GenericOpenVRController.cs:
```diff
@@ -4,13 +4,464 @@
+        #endregion Update data functions
+
+        #region Nested Types
+        private class OpenVRInputMappings
```
Why make a nested class just for const values?

We could probably move this into the main class.

SimonDarksideJ replied:
I made them a sub class to make them easier to refer to :D

SimonDarksideJ replied:
Added a remark at the top of the class to state it's purpose.

StephenHodgson commented on Assets/MixedRealityToolkit/_Core/Utilities/InputMappingAxisUtility.cs:
```diff
@@ -0,0 +1,249 @@
+        /// <summary>
+        /// Used to map AxisType from a useful name to the int value the InputManager wants.
+        /// </summary>
+        public enum MappingAxisType
```
Shouldn't public types be in their own files?

SimonDarksideJ replied:
Normally yet, but I want to keep this code isolated, to make it easier to clean up WHEN we can remove it :D

SimonDarksideJ replied:
Added a remark at the top of the class to state it's purpose.

SimonDarksideJ commented:
Changed Unity InputMapping method to be more generic and inheritable.  GenericOpenVRController is now the base extension for all OpenVR controller types.
Added base implementations for each controller type, including the extensions for the Knuckles controller as an example implementation.

However, haven't found a way with OpenVR to detect controller type.  Will tackle that in another PR.
